File: agent-driven-architecture-main/examples/kotlin/block/inbox/src/main/kotlin/adr/blocks/inbox/Tools.kt
// ── blocks/inbox/tools — the Verb table ────────────────────────────────────
//   noteDrop   Reversible   an input was conflated away or deduped
//   noteFault  Reversible   a turn threw, or a cancel deadline blew
//
// Both are ordinary verbs called through the ordinary path. The consumer never
// touches this block: it reports a spine-shaped `ConsumerEvent`, app/wire maps it to
// these Actions, and they travel resolveAction → gate → fold → commit → signed
// Command like everything else. If an app wires a consumer but NOT this block, the
// Action resolves to a committed `Unhandled` — still on the timeline, still
// observable, never silent.
//
// The decoders guard an OPEN input (6.10): `reason` arrives as a string off the
// wire, so its `when` keeps its else arm and an unknown word fails to decode into a
// committed Unhandled rather than being guessed at.

package adr.blocks.inbox

import adr.contract.InboxCommand
import adr.contract.InboxResult
import adr.contract.InboxResult.DropReason
import adr.spine.pure.Lens
import adr.spine.pure.RawInput
import adr.spine.pure.SourceName
import adr.spine.pure.ToolName
import adr.spine.pure.Verb

val NOTE_DROP = ToolName("noteDrop")
val NOTE_FAULT = ToolName("noteFault")

internal data class NoteDropInput(val source: SourceName, val reason: DropReason, val dropped: Int)

internal data class NoteFaultInput(val source: SourceName, val fault: String)

internal class InboxTools<S>(private val lens: Lens<S, InboxSlice>) {

    fun verbs(): List<Verb<S, *, *>> = listOf(
        Verb.Reversible(
            name = NOTE_DROP,
            describe = "Record that inputs from a source were dropped while the agent was busy.",
            decode = ::decodeNoteDrop,
            run = { input, _ ->
                InboxResult.NoteDrop(NOTE_DROP, input.source, input.reason, input.dropped)
            },
            sign = { r, sig, id -> InboxCommand.NoteDrop(r.tool, sig, id, r.source, r.reason, r.dropped) },
            narrow = { it as? InboxResult.NoteDrop },
        ),
        Verb.Reversible(
            name = NOTE_FAULT,
            describe = "Record that a turn failed or was abandoned, with its cause.",
            decode = ::decodeNoteFault,
            run = { input, _ -> InboxResult.NoteFault(NOTE_FAULT, input.source, input.fault) },
            sign = { r, sig, id -> InboxCommand.NoteFault(r.tool, sig, id, r.source, r.fault) },
            narrow = { it as? InboxResult.NoteFault },
        ),
    )

    private fun decodeNoteDrop(raw: RawInput): NoteDropInput? {
        val source = raw.text("source") ?: return null
        val reason = raw.text("reason")?.let(dropReasonOf::parse) ?: return null
        val dropped = raw.text("dropped")?.toIntOrNull() ?: return null
        return NoteDropInput(SourceName(source), reason, dropped)
    }

    private fun decodeNoteFault(raw: RawInput): NoteFaultInput? {
        val source = raw.text("source") ?: return null
        val fault = raw.text("fault") ?: return null
        return NoteFaultInput(SourceName(source), fault)
    }

    /** OPEN input, guarded: an unrecognised word is a decode failure, never a default. */
    private val dropReasonOf = DropReason.Parser { token ->
        DropReason.entries.firstOrNull { it.name == token }
    }
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/inbox/tools.ts
// ── blocks/inbox/tools — the Verb table ────────────────────────────────────
// Both verbs are `Reversible`: recording that the system shed load changes no
// outside-world state and needs no second principal.
//
// These are ORDINARY verbs, and that is the whole reason the consumer needs no
// new spine machinery to be observable. The consumer emits an Action; the Action
// resolves through the one name→ToolResult map, passes the gate, folds, commits
// and signs. A busy-drop is a decision, so it signs — exactly like 6.8's
// presentation verbs. Wire a consumer without this block and the Action still
// resolves, to a committed `Unhandled`: still on the timeline, never silent.

import type { Verb } from "@adr/spine/pure/verb";
import { reversible } from "@adr/spine/pure/verb";
import type { InferOutput } from "valibot";
import { literal, number, object, string, variant } from "valibot";
import type {
  NoteDropCommand,
  NoteDropResult,
  NoteFaultCommand,
  NoteFaultResult,
} from "./contract";

const dropReason = variant("kind", [
  object({ kind: literal("Conflated"), source: string(), dropped: number() }),
  object({ kind: literal("Duplicate"), source: string(), key: string() }),
]);

export function inboxVerbs<S>(): readonly Verb<S>[] {
  return [
    reversible<S, { reason: InferOutput<typeof dropReason> }, NoteDropResult, NoteDropCommand>({
      name: "noteDrop",
      describe: "Record that an input was dropped — conflated away, or refused as a duplicate.",
      schema: object({ reason: dropReason }),
      run: (input) => ({ outcome: "ok", tool: "noteDrop", reason: input.reason }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "noteDrop",
        sig,
        id,
        reason: result.reason,
      }),
    }),
    reversible<S, { source: string; fault: string }, NoteFaultResult, NoteFaultCommand>({
      name: "noteFault",
      describe: "Record that a turn failed, or that a cancel exceeded its deadline.",
      schema: object({ source: string(), fault: string() }),
      run: (input) => ({
        outcome: "ok",
        tool: "noteFault",
        source: input.source,
        fault: input.fault,
      }),
      sign: (result, sig, id) => ({
        outcome: "ok",
        tool: "noteFault",
        sig,
        id,
        source: result.source,
        fault: result.fault,
      }),
    }),
  ];
}
